backend/src/test/java/com/google/minicrm/servlets/FormsServletTest.java

```diff
@@ -0,0 +1,283 @@
+    response = mock(HttpServletResponse.class);
+  }
+
+  @After
+  public void tearDown() {
+    helper.tearDown();
+  }
+
+  @Test
+  public void formsServlet_getRequest_returnsAllFormsInTimeOrder() throws Exception {
+    slowSeedForms();
+    Form[] returnedForms = getForms();
+    Form[] expectedForms = {form3, form2, form1};
+    assertArrayEquals(expectedForms, returnedForms);
+  }
+
+  @Test
+  public void formsServlet_putRequest_json_successfullyRenamesAndReturns204() throws Exception {
+    seedForms();
+    when(request.getContentType()).thenReturn("application/json;");
+    Reader reader = new StringReader(new FormsPutRequest("2", "newName").toJson());
+    when(request.getReader()).thenReturn(new BufferedReader(reader));
+    formsServlet.doPut(request, response);
+
+    //get the new forms
+    List<Form> returnedForms = Arrays.asList(getForms());
+    form2.setFormName("newName");
+    List<Form> expectedForms = new ArrayList<>(3);
+    expectedForms.add(form1);
+    expectedForms.add(form2);
+    expectedForms.add(form3);
+    //assert all forms are the same regardless of order in list
+    assertTrue(expectedForms.size() == returnedForms.size());
+    assertTrue(expectedForms.containsAll(returnedForms) && expectedForms.containsAll(returnedForms));
+
+    //verify response codes
+    verify(response).setStatus(204);
+    verify(response, never()).sendError(Mockito.anyInt(), Mockito.anyString());
+    verify(response, never()).sendError(Mockito.anyInt());
+  }
+
+  @Test
+  public void formsServlet_putRequest_urlEncoded_successfullyRenamesAndReturns204()
+      throws Exception {
+    seedForms();
+    when(request.getContentType()).thenReturn("application/x-www-form-urlencoded;");
+    when(request.getParameter("formId")).thenReturn("2");
+    when(request.getParameter("formName")).thenReturn("newName");
+    formsServlet.doPut(request, response);
+
+    //get the new forms
+    List<Form> returnedForms = Arrays.asList(getForms());
+    form2.setFormName("newName");
+    List<Form> expectedForms = new ArrayList<>(3);
+    expectedForms.add(form1);
+    expectedForms.add(form2);
+    expectedForms.add(form3);
+    //assert all forms are the same regardless of order in list
+    assertTrue(expectedForms.size() == returnedForms.size());
+    assertTrue(expectedForms.containsAll(returnedForms) && expectedForms.containsAll(returnedForms));
+
+
+    //verify response codes
+    verify(response).setStatus(204);
+    verify(response, never()).sendError(Mockito.anyInt(), Mockito.anyString());
+    verify(response, never()).sendError(Mockito.anyInt());
+  }
+
+  @Test
+  public void formsServlet_putRequest_urlEncoded_throws400WithNoFields() throws Exception {
+    when(request.getContentType()).thenReturn("application/x-www-form-urlencoded;");
+    formsServlet.doPut(request, response);
+
+    //verify the response has an error
+    verify(response).sendError(eq(400), Mockito.anyString()); //we want an error msg
+  }
+
```

Comment:
How indepth should I make these error unit tests? I can think of hundreds of different tests (i.e. throw 400 with no id but a name specified, with a blank id with a valid name, no name valid id, etc.) but do I need to/should I be writing all the different possible combinations that the servlet should return a 400 error or is this around enough?

Reply:
Yeah, I think we should just aim for the most amount of code coverage.Implement as many as you reasonably can while avoiding redundancy. For example, if we already implemented a test for PUT on api/forms where the request has no id but has a name and another request with no name but has an id, then it would be reasonable to assume that no name and no id would not work.